subprojects/base-services/src/test/groovy/org/gradle/internal/classpath/TransformedClassPathTest.groovy

```diff
@@ -0,0 +1,186 @@
+    }
+
+    def "transformed classpath can be mixed with non-transformed one"() {
+        given:
+        TransformedClassPath transformed = transformedClassPath("1.jar": "t1.jar")
+        ClassPath nonTransformed = DefaultClassPath.of(file("2.jar"))
+
+        when:
+        def combined = transformed + nonTransformed
+
+        then:
+        combined.asFiles == [file("1.jar"), file("2.jar")]
+        combined.asTransformedFiles == [file("t1.jar"), file("2.jar")]
+
+        combined.findTransformedJarFor(file("1.jar")) == file("t1.jar")
+        combined.findTransformedJarFor(file("2.jar")) == null
+    }
+
+    def "transformed jars override appended non-transformed ones"() {
+        given:
+        TransformedClassPath transformed = transformedClassPath("1.jar": "t1.jar")
+        ClassPath nonTransformed = DefaultClassPath.of(file("1.jar"))
+
+        when:
+        def combined = transformed + nonTransformed
+
+        then:
+        combined.asFiles == [file("1.jar")]
+        combined.asTransformedFiles == [file("t1.jar")]
+
+        combined.findTransformedJarFor(file("1.jar")) == file("t1.jar")
+    }
+
+    def "transformed classpath can be appended to another transformed"() {
+        given:
+        TransformedClassPath transformed1 = transformedClassPath("1.jar": "t1.jar")
+        TransformedClassPath transformed2 = transformedClassPath("2.jar": "t2.jar")
+
+        when:
+        def combined = transformed1 + transformed2
+
+        then:
+        combined.asFiles == [file("1.jar"), file("2.jar")]
+        combined.asTransformedFiles == [file("t1.jar"), file("t2.jar")]
+
+        combined.findTransformedJarFor(file("1.jar")) == file("t1.jar")
+    }
+
+    def "first transform on the classpath wins"() {
+        given:
+        TransformedClassPath transformed1 = transformedClassPath("1.jar": "t1.jar")
+        TransformedClassPath transformed2 = transformedClassPath("1.jar": "t2.jar")
+
+        when:
+        def combined = transformed1 + transformed2
+
+        then:
+        combined.asFiles == [file("1.jar")]
+        combined.asTransformedFiles == [file("t1.jar")]
+
+        combined.findTransformedJarFor(file("1.jar")) == file("t1.jar")
+    }
+
+    def "transformed classpath can be prepended to non-transformed"() {
+        given:
+        TransformedClassPath transformed = transformedClassPath("1.jar": "t1.jar", "2.jar": "t2.jar")
+        ClassPath nonTransformed = DefaultClassPath.of(file("1.jar"))
+
+        when:
+        def combined = nonTransformed + transformed
+
+        then:
+        combined.asFiles == [file("1.jar"), file("2.jar")]
+        (combined as TransformedClassPath).asTransformedFiles == [file("1.jar"), file("t2.jar")]
+
+        combined.findTransformedJarFor(file("1.jar")) == null
+        combined.findTransformedJarFor(file("2.jar")) == file("t2.jar")
+    }
+
+    def "non-transformed jar on transformed classpath stays non-transformed when another transformation is appended"() {
```

Comment:
❓ What's the rationale for this rule? Intuitively I'd expect a transformed jar to override a non-transformed jar.

Reply:
It follows the behavior that you get in the class loader when you combine transformed and non-transformed JARs with concatenation (or as `DefaultClassPath` instances). AFAIK, the jars in the classpath are searched left-to-right, so the first jar with the matching class wins. If we were to combine `DefaultClassPath` instances, we would get:
```
nonTransformed = DCP(1.jar)
transformed1 = DCP(t2.jar)
transformed2 = DCP(t1.jar)

combined = (nonTransformed + transformed1) + transformed2 => 1.jar:t2.jar:t1.jar
```
As `1.jar` and `t1.jar` contain the same classes, each class loading would end up loading a class from `1.jar`, as it comes first in the combined class path. To keep this behavior, `TransformedClassPath` has to drop the transformation for `1.jar` that comes from `transformed2`.

Reply:
Why not move `t1.jar` to be the transformed counterpart of the early `1.jar` and serve the transformed classes from that earlier point in the classpath?